packages/tpetra/core/src/Tpetra_Reindex_CrsMatrix_def.hpp

```diff
@@ -0,0 +1,125 @@
+#define TPETRA_REINDEX_CRSMATRIX_DEF_HPP
+
+#include <Tpetra_Reindex_CrsMatrix_decl.hpp>
+#include <Tpetra_Vector_decl.hpp>
+#include <Tpetra_Import_decl.hpp>
+
+#include <vector>
+
+/// \file Tpetra_Reindex_CrsMatrix_def.hpp
+/// \brief Definition of the Tpetra::Reindex_CrsMatrix class
+///
+/// If you want to use Tpetra::Reindex_CrsMatrix, include
+/// "Tpetra_Reindex_CrsMatrix.hpp", a file which CMake generates
+/// and installs for you.
+///
+/// If you only want the declaration of Tpetra::Reindex_CrsMatrix,
+/// include "Tpetra_Reindex_CrsMatrix_decl.hpp".
+
+namespace Tpetra {
+
+template <class Scalar,
+          class LocalOrdinal,
+          class GlobalOrdinal,
+          class Node>
+Reindex_CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    Reindex_CrsMatrix(Teuchos::RCP<Map<LocalOrdinal, GlobalOrdinal, Node> const> newRowMap)
+  : ViewTransform<CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node> >()
+  , newRowMap_(newRowMap)
+  , newColMap_(Teuchos::null) {
+  // Nothing to do
+}
+
+template <class Scalar,
+          class LocalOrdinal,
+          class GlobalOrdinal,
+          class Node>
+Reindex_CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::~Reindex_CrsMatrix() {
+  // Nothing to do
+}
+
+template <class Scalar,
+          class LocalOrdinal,
+          class GlobalOrdinal,
+          class Node>
+typename Reindex_CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::NewType
+Reindex_CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>::operator()(OriginalType const& origMatrix) {
+  using cm_t = CrsMatrix<Scalar, LocalOrdinal, GlobalOrdinal, Node>;
+
+  this->origObj_ = origMatrix;
+
+  assert(origMatrix->getRowMap()->getLocalNumElements() == newRowMap_->getLocalNumElements());
+
+  if ((origMatrix->getDomainMap()->getGlobalNumElements() == 0) &&
+      (origMatrix->getRowMap()->getGlobalNumElements() == 0)) {
+    // Construct a zero matrix as a placeholder, don't do reindexing analysis.
+    this->newObj_ = Teuchos::rcp<cm_t>(new cm_t(origMatrix->getRowMap(), origMatrix->getColMap(), 0));
+  } else {
+    using map_t = Map<LocalOrdinal, GlobalOrdinal, Node>;
+    using imp_t = Import<LocalOrdinal, GlobalOrdinal, Node>;
+    using v_t   = Vector<GlobalOrdinal, LocalOrdinal, GlobalOrdinal, Node>;
+
+    // Construct new column map
+    v_t cols(origMatrix->getDomainMap());
+    {
+      size_t origDomainMap_localSize = origMatrix->getDomainMap()->getLocalNumElements();
+      map_t tmpColMap(origMatrix->getDomainMap()->getGlobalNumElements(), origDomainMap_localSize, 0, origMatrix->getDomainMap()->getComm());
+      for (size_t i(0); i < origDomainMap_localSize; ++i) {
+        cols.replaceLocalValue(i, tmpColMap.getGlobalElement(i));
+      }
+    }
+
+    imp_t importer(origMatrix->getDomainMap(), origMatrix->getColMap());
+    v_t newCols(origMatrix->getColMap());
+    newCols.doImport(cols, importer, INSERT, false);
+
+    auto newColsView = newCols.getLocalViewDevice(Tpetra::Access::ReadOnly);
```

Comment:
@eeprude Try moving this view access into the scope of the `parallel_for`, say on line 90.